plugins/org.eclipse.xtend.core/src/org/eclipse/xtend/core/macro/AbstractFileSystemSupport.xtend

```diff
@@ -56,5 +69,220 @@ abstract class AbstractFileSystemSupport implements MutableFileSystemSupport {
 		}
 	}
 	
-	def Path getPath(Resource res);
+	override setContentsAsStream(Path path, InputStream source) {
+		val uri = path.URI
+		if (uri == null)
+			return;
+		
+		if (uri.exists && isFile(uri)) {
+			if (source.markSupported) {
+				if (!hasContentsChanged(source, path.contentsAsStream))
+					return;
+					
+				source.reset
+			}
+		}
+		
+		try {
+			val out = URIConverter.createOutputStream(uri)
+			try {
+				ByteStreams.copy(source, out)
+			} finally {
+				out.close
+			}
+		} catch (IOException exc) {
+			throw new IllegalArgumentException(exc.message, exc)
```

Comment:
Maybe we should document the thrown exceptions and failure conditions also on the intf